Improved Airflow Calculations for BSDF window types

vidanovic opened the pull request:
Pull request overview
---------------------
This version of EnergyPlus introduces improved version of airflow calculations for BSDF window type. It also include several bugfixes.
- Bugfix: Update to frame calculations (correct inside surface temperature passed to frame)
- Effective emissivity calculations for interior shade used incorrect surface emissivity value.
- Homogeneous diffuse shade is now recognized as shade for thermal calculations.
- Edge to center of glass ratio is now taken into account.

Depending on shading construction and surface properties, results can be significantly different from previous version (especially in cases when front and back surface emissivity of shading layer is different.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: 
      - Frame Edge to Center of Glass Ratio is now taken into account (BSDF only). 
      - Other Shading Device type is now included for airflow calculations.
   - Refactoring: No refactoring performed
   - NewFeature: 
      - Update to effective surface temperature and emissivity equations for innermost layer (BSDF only)
      - IR surface exchange algorithm now perform multiple iterations over window surfaces (this is still pending to be implemented and it is not clear why was not implemented in first place, so will need discussion with EnergyPlus team.
   - Performance: No performance change
   - DoNoPublish: /

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check -- verify this
 - [x] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - [x] Documentation changes in place
 - [x] Changed docs build successfully

vidanovic committed:
NFP for this branch.

vidanovic committed:
Fixes for image links

vidanovic committed:
Update AERC_Improvements.md

vidanovic committed:
Update AERC_Improvements.md

vidanovic committed:
Update AERC_Improvements.md

vidanovic committed:
Update AERC_Improvements.md

vidanovic committed:
Update AERC_Improvements.md

vidanovic committed:
Fist pass on merge with what is done in version 8.5

vidanovic committed:
Merge branch '149092157-AERC_updates_to_BSDF' of https://github.com/NREL/EnergyPlus into 149092157-AERC_updates_to_BSDF

vidanovic committed:
Fixed access violation when using non-BSDF with overhang.

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic committed:
Using back surface emissivity when calculating effective emissivity of window with shading device (important for IR calculations)

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic commented:
One note is that I see differences in all complex windows (BSDF) which is expected because of change in some of algorithms.

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic committed:
spaces to tabs

vidanovic committed:
IR iteration loops are now performing window calculations in every loop. This is test for stability, speed and results.

vidanovic committed:
Disable additional iterations for window (to see how many tests will pass)

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic committed:
- Update for frame temperatures for calculations
- IR iterations are now performed.

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic committed:
- IR iterations turned back to single iterations for window
- Fixed two bug that avoid BSDF window calculations when Tsol is set to zero (same bug exists for other type of windows)

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic committed:
- Unit test for new airflow
- Engineering reference for new airflow

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic committed:
Removed unit test (due to undefined behavior of test itself)

vidanovic committed:
Forgot to remove test file from CMakeList

vidanovic committed:
Unit test addition

vidanovic commented:
@Myoldmopar @mjwitte Can and of you guys review this branch? I have put documentation and also made unit tests. Thanks.

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

mjwitte commented:
@vidanovic There are [build warnings](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-3e02d22406adb657a3f5db1537bda07611e0a468-x86_64-Linux-Ubuntu-14.04-gcc-4.8.html).  The source code diffs on github are full of what appear to be unsubstantive indentation and spacing changes - probably your editor's autoformatting run-amok changing things that it shouldn't.  e.g. TarcogShading.hh - everything is indented now within namespace EnergyPlus and space before "&" is gone.  It looks like every source file touched in the pull request has these types of formatting changes.

vidanovic commented:
@mjwitte Do you know any quick way to see where indentation issues are? My way of doing that is to look at github and every line that has been changed. You are right, my editor was set to tab = 2 spaces.

mjwitte commented:
Given  there are so many, it may be easiest to start over with a clean branch and do diffs offline with a tool that lets you ignore whitespace diffs so you can find just the lines where you changed the content.  What editor do you use?

vidanovic commented:
@mjwitte VisualStudio, but I use ReSharper within it.

mjwitte commented:
Ah, you may want to disable auto-formatting or try to get all of the spacing and indentation settings to match up with the current EnergyPlus style.  I know the VS built-in editor will undo auto-formatting with a <ctrl-z> after pasting code or closing a bracket, etc.  I do that a lot, because I can't get all the setting right.

vidanovic commented:
Actually I did set formatting according to wiki (https://github.com/NREL/EnergyPlus/wiki/Coding-Standards-Part-1-Style), but realized lately that tabs are back into the game :)

mjwitte commented:
Sorry - the wiki was more of a proposal for discussion than reality.  In any case, the github diffs will quickly show you when excess changes have crept in.

vidanovic committed:
- Revert back some formatting changes.
- Removed unused includes in some tarcog files

vidanovic commented:
No problem. I actually did changes to resharper and was able to deal with most of things. I had manually to revert empty rows (because resharper format that as well). And another thing is that I could not force it to put namespaces like this:

namespace EnergyPlus {

namespace Something {

but in general, everything else was fine.

One more thing, I was breaking long rows (function calls and if statements). Hopefully that is ok?

vidanovic committed:
- Update to unit test

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic commented:
@mjwitte Can you please review it now? I have fixed all issues and also removed some unnecessary uses. One unit test failed on (https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-a7a099f36b58a3210268d94edbe8100f592d2fae-i386-Windows-7-VisualStudio-14.html) which I cannot repeat locally (maybe force rebuild on remote?). I also see differences in BSDF type of windows, but they are expected.
Thanks.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -1732,19 +1732,21 @@ namespace HeatBalanceSurfaceManager {
-		}
+
+		// This is already initialized in SurfaceWindow structure constructor. Do not do it again here (Simon)
+		// for ( auto & e : SurfaceWindow ) {
```
No, these need to be reset at the beginning of each environment.  `InitThermalAndFluxHistories` is called [here](https://github.com/NREL/EnergyPlus/blob/f4f8b9e466f16ebbb0a72b67b3a3b5677cdcab83/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L497-L501) for each environment.

vidanovic replied:
Will try to compile in debug on ubuntu and see what is going on.
Comments reverted. Initialization is done again.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2867,7 +2869,7 @@ namespace HeatBalanceSurfaceManager {
 							QRadSWwinAbs( IGlass, SurfNum ) += QS( ZoneNum ) * Construct( ConstrNum ).AbsDiffBack( IGlass );
 						}
-					} else if ( ShadeFlag == IntShadeOn || ShadeFlag >= 3 ) {
+					} else if ( ConstrNumSh && ( ShadeFlag == IntShadeOn || ShadeFlag >= 3 ) ) {
```
Should this be ConstrNumSh != 0 ?

vidanovic replied:
Yes

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -5818,7 +5818,10 @@ namespace SolarShading {
+
+						// Do not store in TransDiff because it is not used by BSDF and rest of the code uses it as flag for opaque
+						// surface incorrectly assuming wall heat transfer routines for windows.
+						// Construct( Surface( SurfNum ).Construction ).TransDiff = NomDiffTrans;
```
Did this bug have a broader impact?  If so, we should mention the error and fix in this pull request description (top comment).

vidanovic replied:
Yes, we have this all over the code:
if ( Construct( ConstrNum ).TransDiff <= 0.0 ) { // Opaque surface

Unfortunately, for window with shading devices this can be true statement. I decided not to fix this bug for other window models yet, but we should definitely put it into some bug list.

mjwitte replied:
Issue #6315 posted for the Constuct.TransDiff triggering opaque wall heat transfer problem.

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -6179,7 +6188,10 @@ namespace SolarShading {
+						( ShadeFlag == IntShadeOn || ShadeFlag == ExtShadeOn || ShadeFlag == IntBlindOn ||
+							ShadeFlag == ExtBlindOn || ShadeFlag == BGShadeOn || ShadeFlag == BGBlindOn ||
+							ShadeFlag == ExtScreenOn ) ) {
 						TBmAll = TBmAllShBlSc;
```
I know the style guide mentions wrapping long lines, but we aren't doing that yet per the wiki comment.  @Myoldmopar ?

vidanovic replied:
I will unwrap them for now. I would vote if we can in future wrap them. It is hard to read them in this way.

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -6237,7 +6249,12 @@ namespace SolarShading {
+							SurfaceWindow( SurfNum ).OriginalClass == SurfaceClass_TDD_Diffuser ||
+							Surface( SurfNum ).Class == SurfaceClass_TDD_Dome )
+							continue;
 
```
Same comment about breaking long lines.

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -7969,7 +7986,7 @@ namespace SolarShading {
 			SurfaceWindow( ISurf ).ExtIntShadePrevTS = SurfaceWindow( ISurf ).ShadingFlag;
-			SurfaceWindow( ISurf ).ShadingFlag = NoShade;
+			// SurfaceWindow( ISurf ).ShadingFlag = NoShade;
 			SurfaceWindow( ISurf ).FracTimeShadingDeviceOn = 0.0;
```
Delete?  If there's a good reason to keep this for clarity, then please add a comment to explain.

mjwitte commented on src/EnergyPlus/TARCOGArgs.cc:
```diff
@@ -582,13 +582,17 @@ namespace TARCOGArgs {
 					if ( thick( i ) < SlatThick( i ) ) thick( i ) = SlatThick( i );
 				} else if ( ( ThermalMod == THERM_MOD_ISO15099 ) || ( ThermalMod == THERM_MOD_CSM ) ) {
 					thick( i ) = SlatThick( i );
+					double slatAngRad = SlatAngle( i ) * 2.0 * DataGlobals::Pi / 360.0;
```
No doubles, please use Real64.

mjwitte replied:
@vidanovic This line still needs a Real64 instead of double.

Myoldmopar replied:
@vidanovic This line still needs a Real64 instead of double.

mjwitte commented on src/EnergyPlus/TARCOGArgs.hh:
```diff
@@ -183,6 +183,7 @@ namespace TARCOGArgs {
 		Real64 & Gin,
 		Array1A< Real64 > rir,
 		Array1A< Real64 > vfreevent,
+		Array1A< Real64 > Ah,
```
I realize nearly all of the existing variables lack and endline comment to define what they are, but please add one here to describe what this new variable is, with units if applicable.

mjwitte commented on src/EnergyPlus/TARCOGCommon.cc:
```diff
@@ -57,8 +54,8 @@
-
+namespace EnergyPlus { 
+	
 namespace TARCOGCommon {
```
Minor - some extra whitespace is at the ends of these two line.

mjwitte commented on src/EnergyPlus/TARCOGCommon.cc:
```diff
@@ -98,7 +95,8 @@ namespace TARCOGCommon {
-		if ( ( layertype == VENETBLIND ) || ( layertype == WOVSHADE ) || ( layertype == PERFORATED ) || ( layertype == BSDF ) ) {
+		if ( ( layertype == VENETBLIND ) || ( layertype == WOVSHADE ) || ( layertype == PERFORATED ) ||
+			( layertype == BSDF ) || ( layertype == DIFFSHADE ) ) {
 			IsShadingLayer = true;
```
Another wrapped line.

mjwitte commented on src/EnergyPlus/TARCOGGasses90.cc:
```diff
@@ -144,7 +143,7 @@ namespace TARCOGGasses90 {
 		//REAL(r64), dimension(maxgas, 3) :: xgrho //Autodesk:Unused
-//		static Array2D< Real64 > grho( 3, maxgas ); //Unused
+		//		static Array2D< Real64 > grho( 3, maxgas ); //Unused
 
```
Minor - Do we still need to keep this and above commented lines?

mjwitte commented on src/EnergyPlus/TARCOGGasses90.cc:
```diff
@@ -159,7 +158,7 @@ namespace TARCOGGasses90 {
 
 		//Simon: remove this when assigned properly
-//		grho = 0.0; //Unused
+		//		grho = 0.0; //Unused
```
Delete these two lines? And others later on.  it's not a huge deal if it's useful to keep these around for clarity, except these stick out now because the // got aligned but not there's extra space after the //.

mjwitte commented on src/EnergyPlus/TARCOGGassesParams.hh:
```diff
@@ -52,27 +52,27 @@
 namespace EnergyPlus {
 
-namespace TARCOGGassesParams {
+	namespace TARCOGGassesParams {
```
All changes in this file appear to be an extra tab.

mjwitte commented on src/EnergyPlus/TARCOGMain.cc:
```diff
@@ -707,7 +705,14 @@ namespace TARCOGMain {
+				               ErrorMessage, shgct, tamb, troom, ibc, Atop, Abot, Al, Ar, Ah, SlatThick, SlatWidth, SlatAngle, SlatCond,
+				               SlatSpacing, SlatCurve, vvent, tvent, LayerType, nslice, LaminateA, LaminateB, sumsol, Ra, Nu, ThermalMod,
+				               Debug_mode, ShadeEmisRatioOut, ShadeEmisRatioIn, ShadeHcRatioOut, ShadeHcRatioIn, HcUnshadedOut,
+				               HcUnshadedIn, Keff, ShadeGapKeffConv, SDScalar, SHGCCalc, NumOfIterations, edgeGlCorrFac );
```
Another wrapped line.

mjwitte replied:
Unwrap here.

mjwitte commented on src/EnergyPlus/TARCOGMain.cc:
```diff
@@ -745,7 +750,15 @@ namespace TARCOGMain {
+						               SlatWidth, SlatAngle, SlatCond, SlatSpacing, SlatCurve, vvent, tvent, LayerType, nslice, LaminateA,
+						               LaminateB, sumsol, Ra, Nu, ThermalMod, Debug_mode, ShadeEmisRatioOut, ShadeEmisRatioIn,
+						               ShadeHcRatioOut, ShadeHcRatioIn, HcUnshadedOut, HcUnshadedIn, Keff, ShadeGapKeffConv, SDScalar,
+						               SHGCCalc, NumOfIterations, edgeGlCorrFac );
```
Another wrapped line.

mjwitte replied:
Unwrap.

mjwitte commented on src/EnergyPlus/TARCOGParams.cc:
```diff
@@ -178,6 +178,16 @@ namespace TARCOGParams {
 
 	Real64 const TemperatureQuessDiff( 1.0 ); // in case outside and inside temperatures are identical
 
+	Real64 const C1_VENET( 0.016 );
```
Perhaps a comment here to cite the source of these values?

mjwitte commented on src/EnergyPlus/ThermalISO15099Calc.cc:
```diff
@@ -526,7 +533,18 @@ namespace ThermalISO15099Calc {
+		                       emis, tilt, hin, hout, ibc, SlatThick, SlatWidth, SlatAngle, SlatCond, LayerType, ThermalMod, SDScalar,
+		                       ShadeEmisRatioOut, ShadeEmisRatioIn, ShadeHcRatioOut, ShadeHcRatioIn, Keff, ShadeGapKeffConv, sc, shgc,
+		                       ufactor, flux, LaminateAU, sumsolU, sol0, hint, houtt, trmout, ebsky, ebroom, Gout, Gin, rir, vfreevent,
+		                       Ah, nperr, ErrorMessage );
```
Wrapped line.

mjwitte commented on src/EnergyPlus/ThermalISO15099Calc.cc:
```diff
@@ -526,7 +533,18 @@ namespace ThermalISO15099Calc {
+		}
+
+		updateEffectiveMultipliers( nlayer, width, height, Atop, Abot, Al, Ar, Ah, Atop_eff, Abot_eff, Al_eff,
+		                            Ar_eff, Ah_eff, LayerType, SlatAngle );
```
Wrapped line.

mjwitte commented on src/EnergyPlus/ThermalISO15099Calc.cc:
```diff
@@ -547,7 +565,14 @@ namespace ThermalISO15099Calc {
+			         nperr, ErrorMessage, tamb, troom, ibc, Atop_eff, Abot_eff, Al_eff, Ar_eff, Ah_eff, EffectiveOpenness, vvent, tvent,
+			         LayerType, Ra, Nu, vfreevent, qcgas, qrgas, Ebf, Ebb, Rf, Rb, ShadeEmisRatioOut, ShadeEmisRatioIn,
+			         ShadeHcModifiedOut, ShadeHcModifiedIn, ThermalMod, Debug_mode, AchievedErrorToleranceSolar,
+			         NumOfIterSolar, edgeGlCorrFac );
```
Wrapped.

mjwitte commented on src/EnergyPlus/ThermalISO15099Calc.cc:
```diff
@@ -615,7 +640,14 @@ namespace ThermalISO15099Calc {
+			         ErrorMessage, tamb, troom, ibc, Atop_eff, Abot_eff, Al_eff, Ar_eff, Ah_eff, EffectiveOpenness, vvent, tvent, LayerType,
+			         Ra, Nu, vfreevent, qcgas, qrgas, Ebf, Ebb, Rf, Rb, ShadeEmisRatioOut, ShadeEmisRatioIn,
+			         ShadeHcModifiedOut, ShadeHcModifiedIn, ThermalMod, Debug_mode, AchievedErrorTolerance, NumOfIter,
+			         edgeGlCorrFac );
```
Wrapped.

mjwitte commented on src/EnergyPlus/ThermalISO15099Calc.cc:
```diff
@@ -737,7 +769,16 @@ namespace ThermalISO15099Calc {
+				         Ar_NOSD, Ah_NOSD, EffectiveOpenness_NOSD, vvent_NOSD, tvent_NOSD, LayerType_NOSD, Ra_NOSD, Nu_NOSD, vfreevent_NOSD,
+				         qcgas_NOSD, qrgas_NOSD, Ebf_NOSD, Ebb_NOSD, Rf_NOSD, Rb_NOSD, ShadeEmisRatioOut_NOSD,
+				         ShadeEmisRatioIn_NOSD, ShadeHcModifiedOut_NOSD, ShadeHcModifiedIn_NOSD, ThermalMod, Debug_mode,
+				         AchievedErrorTolerance_NOSD, NumOfIter_NOSD, edgeGlCorrFac );
```
Wrapped.

mjwitte commented on src/EnergyPlus/ThermalISO15099Calc.cc:
```diff
@@ -1159,6 +1206,9 @@ namespace ThermalISO15099Calc {
 			hatter( nlayer, iwd, tout, tind, wso, wsi, VacuumPressure, VacuumMaxGapThickness, ebsky, tamb, ebroom, troom, gap, height, heightt, scon, tilt, theta, Tgap, Radiation, trmout, trmin, iprop, frct, presure, nmix, wght, gcon, gvis, gcp, gama, SupportPillar, PillarSpacing, PillarRadius, hgas, hcgas, hrgas, hcin, hcout, hin, hout, index, ibc, nperr, ErrorMessage, hrin, hrout, Ra, Nu );
 
+			effectiveLayerCond( nlayer, LayerType, scon, thick, iprop, frct, nmix, presure, wght, gcon, gvis, gcp,
+			                    EffectiveOpenness, theta, sconScaled, nperr, ErrorMessage );
```
Wrapped. And several more below.  I'll stop commenting on them individually.

mjwitte commented on src/EnergyPlus/ThermalISO15099Calc.cc:
```diff
@@ -1941,7 +1997,41 @@ namespace ThermalISO15099Calc {
+	                    Array1A< Real64 > const thick, Array2A_int const iprop, Array2A< Real64 > const frct, Array1A_int const nmix,
+	                    Array1A< Real64 > const pressure, Array1A< Real64 > const wght, Array2A< Real64 > const gcon,
+	                    Array2A< Real64 > const gvis, Array2A< Real64 > const gcp, Array1A< Real64 > const EffectiveOpenness,
+	                    Array1< Real64 > & theta, Array1D< Real64 > & sconScaled, int & nperr, std::string & ErrorMessage )
```
Minor - Arguments should be stacked one per line.

mjwitte commented on src/EnergyPlus/WindowComplexManager.cc:
```diff
@@ -3912,27 +3914,30 @@ namespace WindowComplexManager {
+				// SurfaceWindow( SurfNum ).EffShBlindEmiss = EffShBlEmiss;
+				// EffGlEmiss = EpsGlassIR * TauShadeIR / ( 1.0 - RhoGlassIR * RhoShadeIR );
+				// SurfaceWindow( SurfNum ).EffGlassEmiss = EffGlEmiss;
 				//  EffShBlEmiss = InterpSlatAng(SurfaceWindow(SurfNum)%SlatAngThisTS,SurfaceWindow(SurfNum)%MovableSlats, &
```
Delete these?

mjwitte commented on tst/EnergyPlus/unit/WindowLayerEffectiveMultipliers.unit.cc:
```diff
@@ -0,0 +1,165 @@
+		EXPECT_NEAR( Ah_eff( 1 ), 0.008792, 1e-6 );
+
+	}
+}
```
Need a newline at the end of this file.

mjwitte reviewed (changes requested):
Much better - thanks for cleaning up the formatting changes.  I agree the unit test failure is likely a glitch, but the integration test failures [here](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-a7a099f36b58a3210268d94edbe8100f592d2fae-x86_64-Linux-Ubuntu-14.04-gcc-4.8-IntegrationCoverage-Debug.html) are real. These are crashing on linux with a debug build.  If you can't reproduce these with a local windows debug build, then ask @Myoldmopar to run them on linux and send you the error details.  I haven't reviewed doc changes yet.  Here are some code comments.

vidanovic committed:
- Fixed crash on ubuntu debug compilation
- Several formatting issues and code clarification

vidanovic committed:
Fix unused variables in tarcog report routines

vidanovic commented:
@Myoldmopar @mjwitte What are the odds that only one test fails in same way :) And it is not even same test:
From yesterday(9/13/2017)
https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-a7a099f36b58a3210268d94edbe8100f592d2fae-i386-Windows-7-VisualStudio-14.html

and today

https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-9c5ccbaaa209cf6a7a15b4c42d1f0bc9f5c269d8-i386-Windows-7-VisualStudio-14.html

Will try in few hours to upload again, just to see if test(s) is going to fail again.

vidanovic committed:
Description of new airflow coefficients.

vidanovic commented:
@Myoldmopar @mjwitte This time all tests passed (talking of visual studio 14 tests). I think that I know what was the problem. In both previous cases I was pushing branch before previous tests finished and that is what most likely caused  one of test to fail. I would still like this branch to go in before freeze if possible.
Thanks.

mjwitte reviewed (commented):
The diffs for some of the CmplxGlz files are very large.  For example, for CmplxGlz_Daylighting_SouthVB45deg, the [peak heating and cooling loads are now about 40% less than before](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2017-09/f4f8b9e466f16ebbb0a72b67b3a3b5677cdcab83-65ad8b6d9df6c59139aa11587620bb34ade58871/CmplxGlz_Daylighting_SouthVB45deg/x86_64-MacOS-10.9-clang/eplusout.eio.diff.html).  Some of the others also show large diffs.  Is this expected?

(Impressive) Engineering ref additions format well and read OK.

vidanovic committed:
Two lines unwrapped (formatting only)

vidanovic commented:
@mjwitte Yes, some changes are big and changes should be big. We are currently using 8.5 for AERC and I did same changes there (in my local github). As soon as 8.8 is released I will try to run examples here to see how well they match to each other. In general, beside discovering some bugs I have also implemented new airflow. In this moment we are still testing 8.5 version of AERC, so I would not be surprised if we find more bugs. 
Dont know how I missed those two wrappings. Fixed it now and uploaded new code.
Thannks for help Mike. Have a good weekend.

mjwitte commented:
Some mention of the bugs fixed and potential impacts on results would be useful in the Pull Request Overview.  As a user, if I were to read what it says now, I'd expect subtle changes in results.

vidanovic commented:
I did update pull overview (see at very top)

mjwitte commented:
@vidanovic Thanks.  Perfect.
@Myoldmopar I'm finished reviewing here.  Tag - you're it!

Myoldmopar commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -1740,10 +1741,10 @@ namespace HeatBalanceSurfaceManager {
 			e.DividerTempSurfInOld = 23.0;
 			e.DividerTempSurfOut = 23.0;
-
+		
```
Stray whitespace

vidanovic replied:
I can look at both of them. Working on it right now.

Myoldmopar commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -5984,127 +5987,128 @@ namespace SolarShading {
+								// Shade on or switchable glazing
 
-							if ( SunLitFract > 0.0 ) TBmAllShBlSc = POLYF( CosInc, Construct( ConstrNumSh ).TransSolBeamCoef );
+								if ( SunLitFract > 0.0 ) TBmAllShBlSc = POLYF( CosInc, Construct( ConstrNumSh ).TransSolBeamCoef );
```
Stray whitespace

vidanovic replied:
There is "if" statement in line 5990 which goes all the way to 6111. That is why additional additional indentation is here. (Unless I am missing something else?)

Myoldmopar commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -5984,127 +5987,128 @@ namespace SolarShading {
 
+									}
 								}
+
```
What's up with the above indentation?  This little section appears to be indented incorrectly, yet the closing braces are all there?  I don't get what happened.  I'll have to open the file.

vidanovic replied:
Look at above comment.

Myoldmopar commented on src/EnergyPlus/TARCOGDeflection.cc:
```diff
@@ -44,9 +44,6 @@
 
-// C++ Headers
-#include <cmath>
-
```
:+1:

vidanovic replied:
Cheers for CLion that sees these things :)

Myoldmopar commented on src/EnergyPlus/TARCOGMain.cc:
```diff
@@ -707,7 +705,7 @@ namespace TARCOGMain {
 
 			if ( standard == ISO15099 ) {
-				Calc_ISO15099( nlayer, iwd, tout, tind, trmin, wso, wsi, dir, outir, isky, tsky, esky, fclr, VacuumPressure, VacuumMaxGapThickness, CurGap, thick, scon, tir, emis, totsol, tilt, asol, height, heightt, width, presure, iprop, frct, xgcon, xgvis, xgcp, xwght, gama, nmix, SupportPillar, PillarSpacing, PillarRadius, theta, q, qv, ufactor, sc, hflux, hcin, hcout, hrin, hrout, hin, hout, hcgas, hrga ...
+				Calc_ISO15099( nlayer, iwd, tout, tind, trmin, wso, wsi, dir, outir, isky, tsky, esky, fclr, VacuumPressure, VacuumMaxGapThickness, CurGap, thick, scon, tir, emis, totsol, tilt, asol, height, heightt, width, presure, iprop, frct, xgcon, xgvis, xgcp, xwght, gama, nmix, SupportPillar, PillarSpacing, PillarRadius, theta, q, qv, ufactor, sc, hflux, hcin, hcout, hrin, hrout, hin, hout, hcgas, hrga ...
```
This line makes me want to cry a little.  **Sometime** this has **got** to get refactored into a struct so that we only pass like I dunno, less than 20 arguments per function call  :open_mouth: :sob: :upside_down_face: 

I'm partially joking.  But obviously nothing to do for this PR.  Moving on...

vidanovic replied:
Well, I wanted window code refactored for 15 years and finally got green light to do so. Windows-CalcEngine is all about that. Rather then refactoring this part of code, I would rather use WCE...eventually.
BTW, next time you see me remind me to show you conrad code :) (another piece that I am hoping to refactor).

Myoldmopar commented on src/EnergyPlus/TARCOGOutput.cc:
```diff
@@ -1312,19 +1309,19 @@ namespace TARCOGOutput {
 			gio::write( WINCogFile, Format_2031 );
 			gio::write( Format_2020 ) << xgcon( 1, 1 );
-			for ( j = 2; i <= 3; ++j ) {
+			for ( j = 2; j <= 3; ++j ) {
```
Was this just a straight-up bug fix?

vidanovic replied:
Yup, this is bug fix done in WINDOW however, here I noticed it because compiler gave me warning messages. This code is output for WINDOW and it is not used by E+ anyway.

Myoldmopar reviewed (changes requested):
This looks good.  There are a few whitespace things, and I want to look at one section of code where the indent looks really funny to verify it's ok.  The one thing that definitely needs to be fixed is the `double` variable definition still lurking in there.  @vidanovic Can you address that and I'll look at the whitespace things?

vidanovic committed:
Merge branch 'develop' into 149092157-AERC_updates_to_BSDF

vidanovic commented:
@Myoldmopar I have noticed few more double(s) like in this line:
https://github.com/NREL/EnergyPlus/blob/d76c6f0dce2a93593deef94aeb66bad27407fd83/src/EnergyPlus/SolarShading.cc#L3662
That was here before. Let me know if you want it removed?

vidanovic committed:
- Whitespaces removed.
- Missed double to Real64 now fixed.

Myoldmopar commented:
You are right, those aren't really appropriate.  We should have a function called getOurFloatingPoint or something if we really need to coerce a variable into our own typedef'd variable type.  But don't worry about cleaning up all the others that you didn't put in.

Myoldmopar commented:
@vidanovic Looks like you have plenty of blessing on this.  I'll let CI finish up, but if no red flags then I'll merge it.  Thanks for all the effort here.

vidanovic commented:
@Myoldmopar @mjwitte Thank you guys for patience. I did a lot of mess but at least I did learn something.
